Firestore/core/test/firebase/firestore/model/document_key_test.cc

```diff
@@ -0,0 +1,148 @@
+  DocumentKey key({"rooms", "firestore", "messages", "1"});
+  const std::string path_string = "rooms/firestore/messages/1";
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+
+  auto copied = key;
+  EXPECT_EQ(path_string, copied.path().CanonicalString());
+  EXPECT_EQ(key, copied);
+
+  const auto moved = std::move(key);
+  EXPECT_EQ(path_string, moved.path().CanonicalString());
+  EXPECT_NE(key, moved);
+  EXPECT_TRUE(key.path().empty());
+
+  // Reassignment.
+
+  key = copied;
+  EXPECT_EQ(copied, key);
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+
+  key = {};
+  EXPECT_TRUE(key.path().empty());
+  key = std::move(copied);
+  EXPECT_NE(copied, key);
+  EXPECT_TRUE(copied.path().empty());
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+}
+
+TEST(DocumentKey, Constructor_StaticFactory) {
+  const auto key_from_segments =
+      DocumentKey::FromSegments({"rooms", "firestore", "messages", "1"});
+  const std::string path_string = "rooms/firestore/messages/1";
+  const auto key_from_string = DocumentKey::FromPathString(path_string);
+  EXPECT_EQ(path_string, key_from_string.path().CanonicalString());
+  EXPECT_EQ(path_string, key_from_segments.path().CanonicalString());
+  EXPECT_EQ(key_from_segments, key_from_string);
+}
+
+TEST(DocumentKey, Constructor_BadArguments) {
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey(ResourcePath{"foo"}), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey(ResourcePath{"foo", "bar", "baz"}), "");
+
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromSegments({"foo"}), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromSegments({"foo", "bar", "baz"}),
+                            "");
+
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString(""), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString("invalid//string"), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString("invalid/key/path"),
+                            "");
+}
+
+TEST(DocumentKey, IsDocumentKey) {
+  EXPECT_TRUE(DocumentKey::IsDocumentKey({}));
+  EXPECT_FALSE(DocumentKey::IsDocumentKey({"foo"}));
+  EXPECT_TRUE(DocumentKey::IsDocumentKey({"foo", "bar"}));
+  EXPECT_FALSE(DocumentKey::IsDocumentKey({"foo", "bar", "baz"}));
+}
+
+TEST(DocumentKey, Comparison) {
+  const DocumentKey abcd({"a", "b", "c", "d"});
+  const DocumentKey abcd_too({"a", "b", "c", "d"});
+  const DocumentKey xyzw({"x", "y", "z", "w"});
+  EXPECT_EQ(abcd, abcd_too);
+  EXPECT_NE(abcd, xyzw);
+
+  const DocumentKey empty;
+  const DocumentKey a({"a", "a"});
+  const DocumentKey b({"b", "b"});
+  const DocumentKey ab({"a", "a", "b", "b"});
+
+  EXPECT_TRUE(empty < a);
+  EXPECT_TRUE(empty <= a);
+  EXPECT_TRUE(a > empty);
+  EXPECT_TRUE(a >= empty);
+
+  EXPECT_FALSE(a < a);
+  EXPECT_TRUE(a <= a);
+  EXPECT_FALSE(a > a);
+  EXPECT_TRUE(a >= a);
+
+  EXPECT_TRUE(a < b);
+  EXPECT_TRUE(a <= b);
+  EXPECT_TRUE(b > a);
+  EXPECT_TRUE(b >= a);
+
+  EXPECT_TRUE(a < ab);
+  EXPECT_TRUE(a <= ab);
+  EXPECT_TRUE(ab > a);
+  EXPECT_TRUE(ab >= a);
+}
+
```

Comment:
Consider adding tests for [in]equality

EXPECT_TRUE(a == a);
EXPECT_FALSE(a != a);
etc.

Reply:
Done. Added just these two, though, to make sure comparison to self is sensible. `EXPECT_EQ/NE` macros exercise those operators anyway.